File: snippets/design-patterns/structural/flyweight.go
// --- Flyweight Pattern ---
// Use Case: Minimize memory usage by sharing objects with similar properties.
// Example: Caching repetitive data like fonts, icons, or database records.
// Description: The Flyweight Pattern uses shared objects to support large numbers of fine-grained objects efficiently.

package main

import "fmt"

type Flyweight struct {
	sharedState string
}

func (f *Flyweight) Operation(uniqueState string) {
	fmt.Printf("Shared: %s, Unique: %s\n", f.sharedState, uniqueState)
}

type FlyweightFactory struct {
	flyweights map[string]*Flyweight
}

func NewFlyweightFactory() *FlyweightFactory {
	return &FlyweightFactory{flyweights: make(map[string]*Flyweight)}
}

func (f *FlyweightFactory) GetFlyweight(sharedState string) *Flyweight {
	if _, exists := f.flyweights[sharedState]; !exists {
		f.flyweights[sharedState] = &Flyweight{sharedState: sharedState}
	}
	return f.flyweights[sharedState]
}

func main() {
	factory := NewFlyweightFactory()
	flyweight1 := factory.GetFlyweight("shared1")
	flyweight2 := factory.GetFlyweight("shared1")

	flyweight1.Operation("uniqueA") // Output: Shared: shared1, Unique: uniqueA
	flyweight2.Operation("uniqueB") // Output: Shared: shared1, Unique: uniqueB
}

File: snippets/design-patterns/structural/flyweight.py
'''
--- Flyweight Pattern ---
Use Case: Minimize memory usage by sharing objects with similar properties.
Example: Caching icons, fonts, or large sets of similar objects.
'''

class Flyweight:
    def __init__(self, shared_state):
        self.shared_state = shared_state

    def operation(self, unique_state):
        print(f"Shared: {self.shared_state}, Unique: {unique_state}")

class FlyweightFactory:
    _flyweights = {}

    @staticmethod
    def get_flyweight(shared_state):
        if shared_state not in FlyweightFactory._flyweights:
            FlyweightFactory._flyweights[shared_state] = Flyweight(shared_state)
        return FlyweightFactory._flyweights[shared_state]

# Example Usage
factory = FlyweightFactory()
flyweight1 = factory.get_flyweight("shared1")
flyweight2 = factory.get_flyweight("shared1")

flyweight1.operation("uniqueA")
flyweight2.operation("uniqueB")
# Output:
# Shared: shared1, Unique: uniqueA
# Shared: shared1, Unique: uniqueB
